src/OpenLoco/src/World/CompanyManager.cpp

```diff
@@ -617,6 +617,67 @@ namespace OpenLoco::CompanyManager
         company->expenditures[0][static_cast<uint8_t>(type)] -= payment;
     }
 
+    // 0x0046DD06
+    // id : updatingCompanyId global var
+    // payment : ebp
+    // return : ebp == 0x80000000 for false
+    bool ensureCompanyFunding(const CompanyId id, const currency32_t payment)
+    {
+        if (payment <= 0)
+        {
+            return true;
+        }
+        if (isEditorMode())
+        {
+            return true;
+        }
+        if (id == CompanyId::neutral)
+        {
+            return true;
+        }
+
+        auto* company = get(id);
+        if (isPlayerCompany(id))
+        {
+            if ((company->challengeFlags & CompanyFlags::bankrupt) != CompanyFlags::none)
+            {
+                GameCommands::setErrorText(StringIds::company_is_bankrupt);
+                return false;
+            }
+            if (company->cash < payment)
+            {
+                FormatArguments::common(payment);
+                GameCommands::setErrorText(StringIds::not_enough_cash_requires_currency32);
+                return false;
+            }
+            return true;
+        }
+        else
+        {
+            if (company->cash < payment)
+            {
+                const auto maxLoan = Economy::getInflationAdjustedCost(CompanyManager::getMaxLoanSize(), 0, 8);
+                auto additionalFunds = 0;
+
+                while (company->cash + additionalFunds < payment)
+                {
+                    if (company->currentLoan + additionalFunds > maxLoan)
+                    {
+                        FormatArguments::common(payment);
+                        GameCommands::setErrorText(StringIds::not_enough_cash_requires_currency32);
+                        return false;
+                    }
+                    additionalFunds += 1000;
+                }
```

Comment:
Do we really need the loop here? I think we can just calculate the difference and align it back to 1000 and see if its more than its allowed.